IpmiFeaturePkg/GenericIpmi/Pei/PeiGenericIpmi.c

```diff
@@ -54,85 +54,125 @@ PeimIpmiInterfaceInit (
+    DEBUG ((DEBUG_INFO, "[IPMI] IpmiTimeoutPeriod = 0x%x\n", mIpmiInstance->IpmiTimeoutPeriod));
+
+    //
+    // Initialize IPMI IO Base.
+    //
+    mIpmiInstance->Signature                       = SM_IPMI_BMC_SIGNATURE;
+    mIpmiInstance->SlaveAddress                    = BMC_SLAVE_ADDRESS;
+    mIpmiInstance->BmcStatus                       = BMC_NOTREADY;
+    mIpmiInstance->IpmiTransport.IpmiSubmitCommand = IpmiSendCommand;
+    mIpmiInstance->IpmiTransport.GetBmcStatus      = IpmiGetBmcStatus;
+
+    //
+    // Initialize the Ppi descriptor
+    //
+    mPeiIpmiBmcDataDesc->Flags = EFI_PEI_PPI_DESCRIPTOR_PPI | EFI_PEI_PPI_DESCRIPTOR_TERMINATE_LIST;
+    mPeiIpmiBmcDataDesc->Guid  = &gPeiIpmiTransportPpiGuid;
+    mPeiIpmiBmcDataDesc->Ppi   = &mIpmiInstance->IpmiTransport;
+
+    //
+    // Initialize the transport layer.
+    //
+
+    Status = InitializeIpmiTransportHardware ();
+    if (EFI_ERROR (Status)) {
+      DEBUG ((DEBUG_ERROR, "[IPMI] InitializeIpmiTransportHardware failed - %r!\n", Status));
+      return Status;
+    }
+
+    //
+    // Initialize the BMC state.
+    //
+    Status = IpmiInitializeBmc (mIpmiInstance);
+    if (EFI_ERROR (Status)) {
+      DEBUG ((DEBUG_ERROR, "[IPMI] Failed to initialize BMC state. %r\n", Status));
+    }
+
+    //
+    // Initialize the HOB for the DXE phase.
+    //
+
+    BmcHob = BuildGuidHob (&gIpmiBmcHobGuid, sizeof (*BmcHob));
+    if (BmcHob == NULL) {
+      DEBUG ((DEBUG_ERROR, "[IPMI] Failed to create BMC hob!\n"));
+      return EFI_OUT_OF_RESOURCES;
+    }
+
+    BmcHob->BmcStatus = mIpmiInstance->BmcStatus;
+
+    //
+    // Do not continue initialization if the BMC is in Force Update Mode.
+    //
+    if ((mIpmiInstance->BmcStatus == BMC_UPDATE_IN_PROGRESS) || (mIpmiInstance->BmcStatus == BMC_HARDFAIL)) {
+      return EFI_UNSUPPORTED;
+    }
+
+    //
+    // Just produce PPI
+    //
+    Status = PeiServicesInstallPpi (mPeiIpmiBmcDataDesc);
+    if (EFI_ERROR (Status)) {
+      return Status;
+    }
+  } else if (Status == EFI_ALREADY_STARTED) {
+    // This is the execution of the entrypoint after it was shadowed.
+
+    IPMI_BMC_INSTANCE_DATA  *OldIpmiInstance;
+    EFI_PEI_PPI_DESCRIPTOR  *OldPeiIpmiBmcDataDesc;
+    EFI_PEI_HOB_POINTERS    Hob;
+
+    // Locate the existing PPI
+    Status = PeiServicesLocatePpi (&gPeiIpmiTransportPpiGuid, 0, &OldPeiIpmiBmcDataDesc, (VOID **)&OldIpmiInstance);
+    if (!EFI_ERROR (Status)) {
+      // Find the existing allocation by searching memory pool hobs
+      Hob.Raw = GetHobList ();
+      while ((Hob.Raw = GetNextHob (EFI_HOB_TYPE_MEMORY_POOL, Hob.Raw)) != NULL) {
+        // Calculate the correct pointer to the IPMI instance data
+        // The Pool HOB contains the data immediately after the HOB header
+        OldIpmiInstance = (IPMI_BMC_INSTANCE_DATA *)(Hob.Pool + 1);
+
+        if (OldIpmiInstance->Signature == SM_IPMI_BMC_SIGNATURE) {

```

Comment:
This seems a little more dangerous than using a more formal lookup in the heap like a GUID HOB. You might want to check that the HOB pool length (I think would be `Hob.Pool->Header->HobLength - sizeof (EFI_HOB_MEMORY_POOL)`) is at least the size of `IPMI_BMC_INSTANCE_DATA.Signature` (`UINTN`) before checking the signature value. The HOB length should at least be a byte in size to my understanding, but `UINTN` could obviously be larger.